src/agent/onefuzz-supervisor/src/auth.rs

```diff
@@ -87,26 +87,46 @@ pub struct ClientCredentials {
     client_id: Uuid,
     client_secret: Secret<String>,
     resource: String,
+    multi_tenant_domain: Option<String>,
     tenant: String,
 }
 
 impl ClientCredentials {
-    pub fn new(client_id: Uuid, client_secret: String, resource: String, tenant: String) -> Self {
+    pub fn new(
+        client_id: Uuid,
+        client_secret: String,
+        resource: String,
+        multi_tenant_domain: Option<String>,
+        tenant: String,
+    ) -> Self {
         let client_secret = client_secret.into();
 
         Self {
             client_id,
             client_secret,
             resource,
+            multi_tenant_domain,
             tenant,
         }
     }
 
     pub async fn access_token(&self) -> Result<AccessToken> {
+        let (authority, resource) = if let Some(domain) = &self.multi_tenant_domain {
+            let url = Url::parse(&self.resource.clone())?;
+            let host = url.host_str().unwrap();
```

Comment:
We haven't validated that we have a `host_str()` here, so this `unwrap()` may panic, which will abort the program uninformatively. Instead, we could return early with an `Result::Err` and error message.

The `anyhow::bail!()` macro lets us do this explicitly. Example:
```suggestion
            let host = if let Some(host) = url.host_str() {
                host
            } else {
                anyhow::bail!("resource URL does not have a host string: {}", url);
            }
```
Note that the type of the `else` block is unit, and does not match the first block. Since it is an early return, this is okay.

Alternately, you can use the [`Option::ok_or_else()` method](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else) to translate `Option::Some(x)` to `Result::Ok(x)` and the `None` case to a `Result::Err(..)` variant of your choosing (using a closure, to avoid unnecessary computation). This would look like:
```suggestion
            let host = url
                .host_str()
                .ok_or_else(|| anyhow::format_err!("resource URL does not have a host string: {}", url)))?;
```
Note that we use `anyhow::format_err!()` to create a `Result::Err` value (with no early return), and the use of `?` to express "unwrap or return early at this site".